controllers/stucknodecondition/stuck_node_checker.go

```diff
@@ -0,0 +1,341 @@
+			break
+		}
+	}
+	// sort for stable condition
+	sort.Strings(failureMessages)
+
+	// use first 100 messages to avoid blowing out size
+	if len(failureMessages) > 100 {
+		failureMessages = failureMessages[:100]
+	}
+	statusMessage := strings.Join(failureMessages, "; ")
+	// TODO no need if we take this and wire up a client.
+	logger.Info(statusMessage)
+
+	// use real generated client to set only the StuckNode condition
+
+	return nil
+}
+
+type podIdentifier struct {
+	namespace string
+	name      string
+}
+
+// interestingPodsOnStuckNode returns pod namespace,name tuples with a reason why they are interesting
+// Common values are
+// 1. Pod has very long graceful termination
+// 2. Pod has PDB preventing eviction
+func interestingPodsOnStuckNode(ctx context.Context, kubeClient kubernetes.Interface, node corev1.Node, drainTimeoutDuration time.Duration) (map[podIdentifier]error, error) {
+	allPodsOnNode, err := kubeClient.CoreV1().Pods("").List(ctx, metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + node.Name,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list pods: %w", err)
+	}
+
+	podToReason := make(map[podIdentifier]error)
+	for _, pod := range allPodsOnNode.Items {
+		nodeDrainWillNotWaitForPod := false
+		for _, owner := range pod.OwnerReferences {
+			if owner.Kind == "DaemonSet" {
+				nodeDrainWillNotWaitForPod = true
+			}
+		}
+		if nodeDrainWillNotWaitForPod {
+			continue
+		}
+		if isPodFinishedRunning(pod) {
+			continue
+		}
+
+		podKey := podIdentifier{
+			namespace: pod.Namespace,
+			name:      pod.Name,
+		}
+		switch {
+		case pod.DeletionTimestamp == nil:
+			// check to see if the pod is protected by a PDB that prevents deletion
+			if reasons := checkIfPodIsEvictable(ctx, kubeClient, pod); len(reasons) > 0 {
+				podToReason[podKey] = err
+			}
+		case pod.DeletionTimestamp != nil:
+			// check to see if the pod has a really long grace period
+			if err := checkIfPodHasLongGracePeriod(ctx, pod, drainTimeoutDuration); err != nil {
+				podToReason[podKey] = err
+			}
+		}
+	}
+
+	return podToReason, nil
+}
+
+func checkIfPodHasLongGracePeriod(ctx context.Context, pod corev1.Pod, drainTimeoutDuration time.Duration) error {
+	if pod.Spec.TerminationGracePeriodSeconds == nil {
+		return nil
+	}
+	maxGracePeriod := int64(0.8 * drainTimeoutDuration.Seconds())
+	if *pod.Spec.TerminationGracePeriodSeconds > maxGracePeriod {
+		return fmt.Errorf("termination grace period too large: limit=%vs actual=%vs", maxGracePeriod, *pod.Spec.TerminationGracePeriodSeconds)
+	}
+
+	return nil
+}
+
+// checkIfPodIsEvictable reasons why the pod is not evictable
+func checkIfPodIsEvictable(ctx context.Context, kubeClient kubernetes.Interface, pod corev1.Pod) []error {
+	// terminal pods are not subject to PDBs
+	if canIgnorePDB(pod) {
+		return nil
+	}
+
+	// TODO option on caching, might not matter
+	allPDBsInNamespace, err := kubeClient.PolicyV1().PodDisruptionBudgets(pod.Namespace).List(ctx, metav1.ListOptions{})
```

Comment:
There's likely some overlap between this kind of check and [what the managed-upgrade operator is already doing][1], but I'll leave it up to actual MUO maintainers to decide if/how to integrate this with their existing code.

[1]: https://github.com/openshift/managed-upgrade-operator/blob/14278e8ffc0d3e54f326dbc20501af08a9491fbb/pkg/drain/strategy.go#L101-L122

Reply:
+1 We have something similar as part of [nodekeeper controller](https://github.com/openshift/managed-upgrade-operator/blob/master/controllers/nodekeeper/nodekeeper_controller.go#L125) logic. I am thinking if we can merge these.

Also, we did disable these drain strategy for managenent cluster to identify if there is any issue with the workload which we are running.